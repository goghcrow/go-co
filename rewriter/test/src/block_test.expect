// Code generated by github.com/goghcrow/go-co DO NOT EDIT.
package src

import (
	"testing"

	. "github.com/goghcrow/go-co"
	ʂɘʠ "github.com/goghcrow/go-co/seq"
)

func TestBlockStmt(t *testing.T) {
	g := func() ʂɘʠ.Iterator[int] {
		return ʂɘʠ.Start[int](ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
			return ʂɘʠ.Combine[int](ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
				return ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
					return ʂɘʠ.Bind[int](1, func() ʂɘʠ.Seq[int] {
						return ʂɘʠ.Normal[int]()
					})
				})
			}), ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
				return ʂɘʠ.Bind[int](2, func() ʂɘʠ.Seq[int] {
					return ʂɘʠ.Return[int]()
				})
			}))
		}))

	}
	xs := iter2slice(g())
	assertEqual(t, xs, []int{1, 2})
}

func TestTrivalBlock(t *testing.T) {
	g := func() ʂɘʠ.Iterator[int] {
		return ʂɘʠ.Start[int](ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
			for {
				{
					if true {
						break
					} else {
						continue
					}
				}
			}
			return ʂɘʠ.Bind[int](1, func() ʂɘʠ.Seq[int] {
				return ʂɘʠ.Return[int]()
			})
		}))

	}
	xs := iter2slice(g())
	assertEqual(t, xs, []int{1})
}

func TestYieldBlock(t *testing.T) {
	g := func() ʂɘʠ.Iterator[int] {
		return ʂɘʠ.Start[int](ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
			return ʂɘʠ.Combine[int](ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
				return ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
					i := 0
					return ʂɘʠ.For[int](func() bool {
						return i < 5
					}, func() {
						i++
					}, ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
						return ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
							if true {
								return ʂɘʠ.Bind[int](i, func() ʂɘʠ.Seq[int] {
									return ʂɘʠ.Normal[int]()
								})
							} else {
								return ʂɘʠ.Break[int]()
							}
							return ʂɘʠ.Normal[int]()
						})
					}))
				})
			}), ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
				return ʂɘʠ.Return[int]()
			}))
		}))

	}
	xs := iter2slice(g())
	assertEqual(t, xs, []int{0, 1, 2, 3, 4})
}

func TestYieldBlock2(t *testing.T) {
	f := func() {}
	g := func() ʂɘʠ.Iterator[int] {
		return ʂɘʠ.Start[int](ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
			return ʂɘʠ.Combine[int](ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
				return ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
					i := 0
					return ʂɘʠ.For[int](func() bool {
						return i < 5
					}, func() {
						i++
					}, ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
						return ʂɘʠ.Combine[int](ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
							return ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
								if true {
									return ʂɘʠ.Bind[int](i, func() ʂɘʠ.Seq[int] {
										return ʂɘʠ.Normal[int]()
									})
								} else {
									return ʂɘʠ.Break[int]()
								}
								return ʂɘʠ.Normal[int]()
							})
						}), ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
							return ʂɘʠ.Bind[int](1, func() ʂɘʠ.Seq[int] {

								f()
								return ʂɘʠ.Normal[int]()
							})
						}))
					}))
				})
			}), ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
				return ʂɘʠ.Return[int]()
			}))
		}))

	}
	xs := iter2slice(g())
	assertEqual(t, xs, []int{0, 1, 1, 1, 2, 1, 3, 1, 4, 1})
}
