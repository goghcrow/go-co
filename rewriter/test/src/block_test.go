package src

import (
	"testing"

	. "github.com/goghcrow/go-co"
)

func TestBlockStmt(t *testing.T) {
	g := func() Iter[int] {
		{
			Yield(1)
		}
		Yield(2)
		return nil
	}
	xs := iter2slice(g())
	assertEqual(t, xs, []int{1, 2})
}

func TestTrivalBlock(t *testing.T) {
	g := func() Iter[int] {
		for {
			{
				if true {
					break
				} else {
					continue
				}
			}
		}
		Yield(1)
		return nil
	}
	xs := iter2slice(g())
	assertEqual(t, xs, []int{1})
}

func TestYieldBlock(t *testing.T) {
	g := func() Iter[int] {
		for i := 0; i < 5; i++ {
			{
				if true {
					Yield(i)
				} else {
					break
				}
			}
		}
		return nil
	}
	xs := iter2slice(g())
	assertEqual(t, xs, []int{0, 1, 2, 3, 4})
}

func TestYieldBlock2(t *testing.T) {
	f := func() {}
	g := func() Iter[int] {
		for i := 0; i < 5; i++ {
			{
				if true {
					Yield(i)
				} else {
					break
				}
			}
			Yield(1)
			f()
		}
		return nil
	}
	xs := iter2slice(g())
	assertEqual(t, xs, []int{0, 1, 1, 1, 2, 1, 3, 1, 4, 1})
}
