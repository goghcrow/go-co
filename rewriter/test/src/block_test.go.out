//go:build !co

// Code generated by github.com/goghcrow/go-co DO NOT EDIT.
package src

import (
	ʂɘʠ "github.com/goghcrow/go-co/seq"
	"testing"
)

func TestBlockStmt(t *testing.T) {
	g := func() ʂɘʠ.Iterator[int] {
		return ʂɘʠ.Start[int](
			ʂɘʠ.Combine[int](

				ʂɘʠ.Bind[int](1,
					ʂɘʠ.Normal[int],
				),

				ʂɘʠ.Bind[int](2,
					ʂɘʠ.Return[int],
				),
			),
		)

	}
	xs := iter2slice(g())
	assertEqual(t, xs, []int{1, 2})
}

func TestTrivalBlock(t *testing.T) {
	g := func() ʂɘʠ.Iterator[int] {
		return ʂɘʠ.Start[int](ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
			for {
				{
					if true {
						break
					} else {
						continue
					}
				}
			}
			return ʂɘʠ.Bind[int](1,
				ʂɘʠ.Return[int],
			)
		}))

	}
	xs := iter2slice(g())
	assertEqual(t, xs, []int{1})
}

func TestYieldBlock(t *testing.T) {
	g := func() ʂɘʠ.Iterator[int] {
		return ʂɘʠ.Start[int](
			ʂɘʠ.Combine[int](
				ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
					i := 0
					return ʂɘʠ.For[int](func() bool {
						return i < 5
					}, func() {
						i++
					},
						ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
							if true {
								return ʂɘʠ.Bind[int](i,
									ʂɘʠ.Normal[int],
								)
							} else {
								return ʂɘʠ.Break[int]()
							}
						}),
					)
				}),

				ʂɘʠ.Return[int](),
			),
		)

	}
	xs := iter2slice(g())
	assertEqual(t, xs, []int{0, 1, 2, 3, 4})
}

func TestYieldBlock2(t *testing.T) {
	f := func() {}
	g := func() ʂɘʠ.Iterator[int] {
		return ʂɘʠ.Start[int](
			ʂɘʠ.Combine[int](
				ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
					i := 0
					return ʂɘʠ.For[int](func() bool {
						return i < 5
					}, func() {
						i++
					},
						ʂɘʠ.Combine[int](
							ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
								if true {
									return ʂɘʠ.Bind[int](i,
										ʂɘʠ.Normal[int],
									)
								} else {
									return ʂɘʠ.Break[int]()
								}
							}),

							ʂɘʠ.Bind[int](1, func() ʂɘʠ.Seq[int] {

								f()
								return ʂɘʠ.Normal[int]()
							}),
						),
					)
				}),

				ʂɘʠ.Return[int](),
			),
		)

	}
	xs := iter2slice(g())
	assertEqual(t, xs, []int{0, 1, 1, 1, 2, 1, 3, 1, 4, 1})
}
